My project pizza 1

turgunbek opened the pull request:
Нужно было реализовать программу для имитации заказа Пиццы с возможностью выбора из командной строки (консоли), используя ООП и ФП, и по возможности - добавлением тестов.
Все требования выполнены. Программа запускается корректно, согласно смыслу задачи. Написаны основные тесты.

turgunbek committed:
Добавлен основной файл, реализующий решение задачи.

turgunbek committed:
Добавлен файл 'tests.py' с тестами и файл с результатами тестов 'test_results.txt'

turgunbek committed:
Добавлен файл с результатами запуска программы (из терминала с возможностью прописывать опции)

turgunbek committed:
Update README.md

turgunbek committed:
Update README.md

turgunbek committed:
Update README.md

turgunbek committed:
Update README.md

pulysak commented on cli.py:
```diff
@@ -0,0 +1,127 @@
+
+
+class Margherita(Pizza):
+    """Класс наследуется от базового класса Pizza.
```
https://peps.python.org/pep-0257/
Вот тут можно почитать про правила написания докстрингов.
Однострочник должен выглядеть вот так.

"""Класс наследуется от базового класса Pizza."""
Но это все придирки

pulysak commented on cli.py:
```diff
@@ -0,0 +1,127 @@
+    Содержит атрибуты об имени, эмоджи, ингридиентах и размере пиццы.
+    Содержит метод dict() для получения списка ингридиентов.
+    """
+    def __init__(self, name, emoji, ingredients, size='L'):
```
Было бы круто(но не обязательно) использовать https://docs.python.org/3/library/enum.html для размеров пицц

pulysak commented on cli.py:
```diff
@@ -0,0 +1,127 @@
+        self.size = size
+
+    def dict(self) -> str:
+        return ', '.join(self.ingredients)
```
По заданию нужно отдавать словарь) ну и название метода тоже говорит об этом

Можно просто сделать ключ с названием пиццы, а значение ингридиенты или придумать что-то еще

turgunbek replied:
Сделал. Теперь формально возвращает словарь)

pulysak commented on README.md:
Хорошо бы разбить файл на несколько. Отдельно выделить код с пиццой, отдельно cli и например декоратор

turgunbek replied:
Сделано. Спасибо за этот комментарий. Совсем забыл про модульность.

pulysak commented on cli.py:
```diff
@@ -0,0 +1,127 @@
+@click.argument('pizza_type', type=click.STRING)
+def order(delivery: bool, size: str, pizza_type: str) -> None:
+    """Готовит пиццу и доставляет её (или оставляет для самовывоза)"""
+    if pizza_type.lower() == 'margherita':
```
Такой набор иф-ов иногда заменяют словарем. Можно сделать так:
```python
    menu = {
        'margherita': Margherita(),
        'pepperoni': Pepperoni(),
        'hawaiian': Hawaiian(),
    }
    
    pizza = menu.get(pizza_type)
    if not pizza:
        click.echo(f'Пицца с именем {pizza_type} не найдена в меню.')
        return
    ...

Можно даже подумать про вынос menu в константы
```

pulysak commented on tests.py:
```diff
@@ -0,0 +1,115 @@
+    """
+    pickup()
+    captured = capsys.readouterr()
+    assert 'Самовывоз' in captured.out
```
в этих тестах можно запатчить randint и проверять строку целиком

turgunbek replied:
Сделано.

pulysak commented on tests.py:
```diff
@@ -0,0 +1,115 @@
+    """
+    runner = CliRunner()
+
+    # Тест 1: Проверяем создание объекта Margherita
```
Если в тестах есть повторяющийся паттерн, то можно написать его через https://docs.pytest.org/en/7.1.x/how-to/parametrize.html#pytest-mark-parametrize-parametrizing-test-functions

turgunbek replied:
Сделано.

pulysak reviewed (commented):
В целом все ок. Сниму 3 балла за отсутсвие разбивки по файлам. Можно поправить, тогда поднимем баллы.
Остальное можно поправить по желанию

turgunbek committed:
Исправлено в соответствии с замечаниями:
1. Программа разбита на модули (pizza.py, decorators.py, cli.py).
2. Переработан метод dict() в соответствии с заданием - чтобы возвращал словарь.
3. Переработаны тесты: запатчен randint, использованы параметрические тесты.
4. Сделаны небольшие улучшения в коде.

turgunbek committed:
Merge branch 'my-project-pizza-1' of https://github.com/turgunbek/project_pizza into my-project-pizza-1

turgunbek committed:
Обновил файл с выводом результатов тестов.

turgunbek commented:
Сделал коммит, все замечания (кроме enum для размеров пицц - может попозже посижу ещё поразбираюсь) учтены, и переделал программу в соответствии с ними.
